include/crc16.h

```diff
@@ -44,10 +44,7 @@ u16_t crc16(const u8_t *src, size_t len, u16_t polynomial,
  *
  * @return The computed CRC16 value
  */
-static inline u16_t crc16_ccitt(const u8_t *src, size_t len)
-{
-	return crc16(src, len, 0x1021, 0xffff, true);
-}
+u16_t crc16_ccitt(const u8_t *src, size_t len);
```

Comment:
Speeding up the computation is awsome, but I have a few comments:
1) why is the improvement made only for the ccitt variant and not the generic one? All other variants will now be slower without any good reason.
2) test cases must be added, else the next person (or even this implementation) cannot be known to compute correct values
3) this implementation works only on continuous buffers, which is not the case usually. This is why bool pad was introduced later on.

Reply:
Hi @pfl.  Thanks for the review!

re: optimisation, I've added test vectors and a byte-at-a-time version for crc16_ansi.

re: test cases, there's already vectors in tests/unit/drivers/crc/main.c.  I've added vectors for crc16_ansi.

re: pad, the bitwise version also only supports a single, contiguous buffer.  'pad' is something different - it causes an extra two bytes of zeros to be fed through the CRC function as required by the CCITT spec.  The bytewise version achieves the same by using a different seed.  See http://srecord.sourceforge.net/crc16-ccitt.html for more.

Reply:
Hi @tbursztyka see my reply to @pfl above.

Reply:
Thanks for the tests!
For 1) above I actually meant that there would be one function that computes the CRC, and then the variants would call this function with their initial values and polynomial. But something smart needs to be done along the way for XOR, since it now is not as trivial as before when it was done bit by bit.
About the pad value: yes, the input buffer is continuous. But having the pad added or not allowed the computation to continue thus enabling the computation to be done over N continuous buffers, see commit 5d8eadd7b1aa19b1b33ce88f2a06ca57f6222c10 where it was introduced. This is needed by some, so that should still be supported.

Reply:
Ah, sorry, in my earlier reply I got 'bitwise' and 'bytewise' mixed up in my head.

re: crc16, sorry, I don't know a way of doing a fast, runtime configured crc16.  The bytewise version is 10x faster as the loop is unrolled and known at compile time.  The best options I can think of are:

1. Change crc16 to call the specialised ones when the parameters are the right values.  This would increase the code size as all versions would get linked in.

2. Switch to a table driven method as the table can be filled in on first use.  This would need a cache and 256 x 2 bytes of RAM for the table itself.

re: the use case, you can chain the bytewise versions by changing the API to allow the initial CRC to be passed in.  Something like:

`u16_t crc16_ccitt_seeded(u16_t initial_crc, void *buf, int len, bool final)`

I'm not sure that 'final' is needed - I'm pretty sure that if you use 0x1d01 as a seed instead of 0xffff then you can skip the final 16 bits of zeros.

Reply:
I looked further, and the only user of crc16 is the nffs which is originally from mynewt.  They use the chained CRC method:

https://github.com/apache/mynewt-core/blob/master/fs/nffs/src/nffs_crc.c#L110

with the following API:

https://github.com/apache/mynewt-core/blob/master/util/crc/src/crc16.c#L69